### ISSUE TITLE
Making a Configuration Copy throws ClassCastException if Type other than DefaultDependencyConstraint among Constraints

### ISSUE THREAD

**Issue by mroeppis** [gradle/gradle#17179]
The multi-module project I'm working on uses [version alignment](https://docs.gradle.org/current/userguide/dependency_version_alignment.html). As a result my BOM module holds configurations with submodule constraints. Since I'm working on a custom plugin I need to create editable copies for all configurations in my project. When trying to make a copy of a configuration with submodule constraints the following exception is raised: 

```groovy
"java.lang.ClassCastException: class org.gradle.api.internal.artifacts.dependencies.DefaultProjectDependencyConstraint cannot be cast to class org.gradle.api.internal.artifacts.dependencies.DefaultDependencyConstraint (org.gradle.api.internal.artifacts.dependencies.DefaultProjectDependencyConstraint and org.gradle.api.internal.artifacts.dependencies.DefaultDependencyConstraint are in unnamed module of loader org.gradle.initialization.MixInLegacyTypesClassLoader @7a5ceedd)"
```

I discovered that `DefaultConfiguration` does the following:

https://github.com/gradle/gradle/blob/7286dcd1d9b46297fd62231cc7fbf17d958604aa/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java#L1141-L1143

In other words, if a `DefaultProjectDependencyConstraint` is among `dependencyConstraints` this won't succeed. From my point of view this behavior can't be intended since the troublesome type is valid among configuration constraints.

Possible solutions:
* remove cast and place `copy()` inside `DependencyConstraint` to have all subtypes override it
* offer `copy(...)` with param `Spec<? super DependencyConstraint> dependencyConstraintSpec` which could be used to filter out constraints not meant to be copied -> still there must be a good reason not to make some constraints copiable.

---

My Gradle version is `6.7.1`.

**Comment by ljacomet** [comment]
`copy` should be moved to `DependencyConstraintInternal` and this cast should use that type.
